### PR TITLE
[#8029] add audit logging for server start

### LINKED ISSUE #8029: [Audit|Backend] Implement FHIR AuditEvent logging service [3]
### Situation Description

For the general purpose of audit logging (#7904) we need to be able to log access to the SORMAS backend with details on what exactly has been accessed, without including sensitive data (e.g. names).

This issue is the first step. It defines how the logged data is represented and provides a service that allows to log audit events.

#### AuditEvent

For the logging output format, we are using the [FHIR R4 AuditEvent resource](https://www.hl7.org/fhir/auditevent.html#tabs-json), which is based on the IHE ATNA audit profile. FHIR is a well-known and established standard for exchange of medical related data. 

A `FHIR AuditEvent` resource has a specified [JSON representation](https://www.hl7.org/fhir/auditevent-example.json.html) which is compact and easy to digest by log processing systems like Loki or ELK.

An open source library (`hapifhir`) that provides the needed `AuditEvent` java classes is available here: [Github](https://github.com/hapifhir/org.hl7.fhir.core/tree/master/org.hl7.fhir.r4) [Maven](https://mvnrepository.com/artifact/ca.uhn.hapi.fhir/org.hl7.fhir.r4/5.6.35)

These are the most important fields in the AuditEvent class:

|Content|Description|Field in `AuditEvent` resource|
|-----|------|-------|
|Access type| Create/Read/Update/Delete/Execute| `action`
|timestamp| Timestamp of the event| `recorded`
|actor|Functional instance causing the event (e.g., user). Users should be identified with a human readable name besides using the UUID| `agent`
|Executing instance| Identifaction of the system(component) generating the audit event| `source.site`
|Executing class|Name of the executing class, including method name, parameters, and return values| `source.observer`
|Activity/Event| Description of the activity with unique reference(uuid) w.r.t. the accesses data| `entitiy`

Additional, there is the field `**type**`, which will be populated according to the following table.

| Code | System | Name | Description |
| ---- | ------ | ---- | ----------- |
| 110100 | http://dicom.nema.org/resources/ontology/DCM | Application Activity | Start, Stop |
| 110106 | http://dicom.nema.org/resources/ontology/DCM | Export | Database level export |
| 110112 | http://dicom.nema.org/resources/ontology/DCM | Query | A request for multiple entities |
| 110110 | http://dicom.nema.org/resources/ontology/DCM | Patient Record | Use for read/write/delete operations on patient related entities (case, contact, symptoms, samples, ...) |
| object | http://terminology.hl7.org/CodeSystem/audit-event-type | An Operation on other Objects | Use for read/write/delete operations on all other entities, most importantly users, infrastructure, configuration and tasks |

#### AuditLogger / Implementation

- [x] create an `AuditLogger` class following the Singleton pattern (may be changed later, e.g. provinding one logger per class, similar to what Logback does).
- [x] Use the `AuditEvent` class of `HapiFhir` [Github](https://github.com/hapifhir/org.hl7.fhir.core/tree/master/org.hl7.fhir.r4) [Maven](https://mvnrepository.com/artifact/ca.uhn.hapi.fhir/org.hl7.fhir.r4/5.6.35). Make sure to provide this as a domain library and exclude optional dependencies.
- [x] Add a private method that takes a FHIR `AuditEvent` class, serializes it to json and writes the json into the default log (using slf4j Logger, level INFO). Later this will be replaced with a different log sink mechanism (see  #8030)
- [x] Add a first `logApplicationStart` method that is called from `StartupShutdownService`. It creates an `AuditEvent` and fills the following fields:
  * `type, subtype` (110100 and 110120) - HAP FHIR does not seem to provide this, so we should create our own enums to provide the Coding
  * `action`: E
  * `recorded`: current timestamp
  * `agent`: **Discuss** Is this needed for application start? Probably not
  * `source.site:` URL of the server - **discuss* how to provide this
  * `source.type`: https://www.hl7.org/fhir/codesystem-audit-source-type.html#audit-source-type-4
- [x] Same for `logApplicationStop`

More methods for the other log types will be implemented in follow-up issues (see #8025 for complete list).

### Possible Alternatives

Use the IPF commons audit library that provided an AuditMessage and builders for the different message types: https://oehf.github.io/ipf-docs/docs/audit/

It also allows to provide a `serializationStrategy` to control how the audit event is represented. Using
the FHIR R4 JSON [serializer](https://github.com/oehf/ipf/blob/0b104be3cb4756956e2a8cb8d19d9f0f23a4792c/commons/ihe/fhir/r4/core/src/main/java/org/openehealth/ipf/commons/ihe/fhir/support/audit/marshal/FhirAuditJsonSerializationStrategy.java) from the `org.openehealth.ipf.commons.ihe.fhir.support.audit.marshal` package will provide us with the desired result.

Unfortunately the AudtiMessage does not seem to provide everything that is needed, e.g. the source data does allow us to provide the observer information needed to log the executing class and method details (https://github.com/oehf/ipf/blob/master/commons/audit/src/main/java/org/openehealth/ipf/commons/audit/model/AuditSourceIdentificationType.java).

### Additional Information

Example of a log message for type "Query":

<details>

```
{ 
  "id": "05939825-617c-4bef-b5be-eda37bc61518", 
  "type": { 
    "system": http://dicom.nema.org/resources/ontology/DCM, 
    "code": "110112", 
    "display": "Query" 
  }, 
  "action": "R", 
  "recorded": "2021-12-08T23:41:23Z", 
  "outcome": "0", 
  "agent": [ 
    { 
      "type": { 
        "coding": [ 
          { 
            "system": http://terminology.hl7.org/CodeSystem/extra-security-role-type, 
            "code": "humanuser", 
            "display": "human user" 
          } 
        ] 
      }, 
      "who": { 
        "identifier": { 
          "value": "OHALA5-VLODZTQ-GKQ7V3-FDSMINPM" 
        } 
      }, 
      "name": "Steff Surveillance", 
      "requestor": true 
    }, 
  ], 
  "source": { 
    "site": "musterstadt.sormas-oegd.de", 
    "observer": { 
      "identifier": { 
        "value": "de.symeda.sormas.api.caze.CaseFacade.getIndexList" 
      } 
    }, 
    "type": [ 
      { 
        "system": http://terminology.hl7.org/CodeSystem/security-source-type, 
        "code": "4", 
        "display": "Application Server" 
      } 
    ] 
  }, 
  "entity": [ 
    { 
      "query": "aHR0cDovL2ZoaXItZGV2LmhlYWx0aGludGVyc2VjdGlvbnMuY29tLmF1L29wZW4vRW5jb3VudGVyP3BhcnRpY2lwYW50PTEz" 
	}, 
    { 
      "what": { 

        "identifier": { 
            "value": "urn:de.sormas.symeda.api.caze.CaseIndexDto:XEWVWM-DZVFUF-KTHZII-G2EXSC3I" 
         } 
      } 
    }, 
    { 
      "what": { 
        "identifier": { 
            "value": "urn:de.sormas.symeda.api.caze.CaseIndexDto:SQF6TW-FP2RFM-4FPDS5-WZLSSI4I" 
         } 
      } 
    },    { 
      "what": { 
        "identifier": { 
            "value": "urn:de.sormas.symeda.api.caze.CaseIndexDto: RC7ITS-OEYALC-K7NV2X-CPSPKPJQ " 
         } 
      } 
    } 
  ] 
} 
```

</details>

### PR DESCRIPTION
**Pull request by JonasCir**
Fixes #8029 

```json
{
  "resourceType": "AuditEvent",
  "type": {
    "system": "https://hl7.org/fhir/R4/valueset-audit-event-type.html",
    "code": "110100",
    "display": "Application Activity"
  },
  "subtype": [
    {
      "system": "https://hl7.org/fhir/R4/valueset-audit-event-sub-type.html",
      "code": "110120",
      "display": "Application Start"
    }
  ],
  "action": "E",
  "recorded": "2022-03-10T13:21:46.875+01:00",
  "outcome": "0",
  "outcomeDesc": "Application starting",
  "agent": [
    {
      "type": {
        "coding": [
          {
            "system": "https://www.hl7.org/fhir/valueset-participation-role-type.html",
            "code": "110151",
            "display": "Application Launcher"
          }
        ]
      },
      "name": "SYSTEM"
    }
  ],
  "source": {
    "site": "sormas_b",
    "type": [
      {
        "system": "http://terminology.hl7.org/CodeSystem/security-source-type",
        "code": "4",
        "display": "Application Server"
      }
    ]
  },
  "entity": [
    {
      "what": {
        "reference": "StartupShutdownService"
      },
      "type": {
        "system": "http://terminology.hl7.org/CodeSystem/audit-entity-type",
        "code": "2",
        "display": "System Object"
      }
    }
  ]
}
```

### CODE REVIEW

File: sormas-base/pom.xml [hunk 1]
```diff
@@ -584,10 +584,22 @@
 				<version>${slf4j.version}</version>
 				<scope>provided</scope>
 			</dependency>
+			<dependency>
+				<groupId>ca.uhn.hapi.fhir</groupId>
+				<artifactId>hapi-fhir-structures-r4</artifactId>
+				<version>5.7.0</version>
+				<scope>runtime</scope>
```
Comment (MartinWahnschaffe) [inline sormas-base/pom.xml:591]: should be provided and the lib also needs to be added to the serverlibs.pom

File: sormas-base/pom.xml [hunk 2]
```diff
@@ -584,10 +584,22 @@
...
+			</dependency>
 
 			<!-- *** Domain libs END *** -->
 
 			<!-- *** Compile dependencies *** -->
+			<dependency>
```
Comment (MartinWahnschaffe) [inline sormas-base/pom.xml:597]: shouldn't this also be provided as a serverlib in the domain?

**Review by MartinWahnschaffe — CHANGES_REQUESTED** [review]
Already looks quite good!

See my comments on the dependencies.

Why is the source.site such a strange number string in your example?

**Comment by JonasCir** [comment]
@MartinWahnschaffe I don't really know how all this maven dependency magic plays out. With this configuration it just works :tm: .
I tried to provide them as server libs, but then: 1. compilation failed b/c some imports were not available during compile 2. and second I ran into an issue winch already prevents other libs to be deployed as serverlib.

I'm not familiar with this, Is there any particular reason we need to change it? `org.hl7.fhir.r4` must be available at compile time, while `hapi-fhir-structures-r4` is purely a runtime dependency.

**Comment by MartinWahnschaffe** [comment]
Yes, the reason to change is is to make sure the build artifact that needs to be deployed does not become 10MB bigger again. Also it's very likely that the FHIR resource will be needed by the rest, backend and ui archives, so we should not put it into the server three times (included in each archive).

You can have a look at the com.opencsv dependency. If structures is just a runtime depdency you probably don't have to mention it in the backend pom.

On the server URL: We already have someting similar for the app.url property. We can't re-use that one, but it means that we probably already have a mechanism in sormas docker that fills in urls into the property file.

**Comment by JonasCir** [comment]
Hi @MartinWahnschaffe 
thanks for the feedback. I cannot deploy the dependencies as server libs right now due to https://github.com/hzi-braunschweig/SORMAS-Project/issues/8324, which as also affected the dependencies for Keycloak in the past.

My proposed way forward given the time constraints we facing is:
1. Leave the dependencies as they are now
2. Fix #8324 seperately
3. Introduce a new server property for the hostname. (app.url is also set like the rest of the properties, see [here](https://github.com/hzi-braunschweig/SORMAS-Docker/blob/master/sormas/start-server.sh#L236)) 
4. Merge this to unblock https://github.com/hzi-braunschweig/SORMAS-Project/issues/8025

File: sormas-base/pom.xml (hunk 1 above)
Comment (JonasCir) [reply sormas-base/pom.xml:591]: https://github.com/hzi-braunschweig/SORMAS-Project/issues/8324

File: sormas-base/pom.xml (hunk 2 above)
Comment (JonasCir) [reply sormas-base/pom.xml:597]: https://github.com/hzi-braunschweig/SORMAS-Project/issues/8324

**Review by MartinWahnschaffe — APPROVED** [review]

**Pull request merged**